File: app/src/main/java/de/eloc/eloc_control_panel/ng2/activities/RegisterActivity.java
package de.eloc.eloc_control_panel.ng2.activities;

import static de.eloc.eloc_control_panel.ng3.activities.ActivityExtensionsKt.hideKeyboard;
import static de.eloc.eloc_control_panel.ng3.activities.ActivityExtensionsKt.open;
import static de.eloc.eloc_control_panel.ng3.activities.ActivityExtensionsKt.showModalAlert;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.view.View;

import androidx.lifecycle.ViewModelProvider;

import de.eloc.eloc_control_panel.R;
import de.eloc.eloc_control_panel.databinding.ActivityRegisterBinding;
import de.eloc.eloc_control_panel.ng3.activities.LoginActivity;
import de.eloc.eloc_control_panel.ng3.activities.ThemableActivity;
import de.eloc.eloc_control_panel.ng3.activities.VerifyEmailActivity;
import de.eloc.eloc_control_panel.ng3.data.UserAccountViewModel;

public class RegisterActivity extends ThemableActivity {
    private ActivityRegisterBinding binding;
    private UserAccountViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityRegisterBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        viewModel = new ViewModelProvider(this).get(UserAccountViewModel.class);

        setListeners();
        setWatchers();
        updateUI(false);
    }

    private void setWatchers() {
        binding.emailAddressTextInput.addTextChangedListener(new TextInputWatcher(binding.emailAddressLayout));
        binding.passwordTextInput.addTextChangedListener(new TextInputWatcher(binding.passwordLayout));
        binding.verifyPasswordTextInput.addTextChangedListener(new TextInputWatcher(binding.verifyPasswordLayout));
    }

    private void setListeners() {
        binding.backButton.setOnClickListener(v -> onBackPressed());
        binding.loginButton.setOnClickListener(v -> onBackPressed());
        binding.registerButton.setOnClickListener(v -> register());
        binding.root.setOnClickListener(v -> hideKeyboard(RegisterActivity.this));
    }

    @Override
    public void onBackPressed() {
        Intent intent = new Intent(this, LoginActivity.class);
        startActivity(intent);
        finish();
    }

    private void register() {
        hideKeyboard(this);
        binding.emailAddressLayout.setError(null);
        binding.passwordLayout.setError(null);
        binding.verifyPasswordLayout.setError(null);
        String emailAddress = "";
        Editable editable = binding.emailAddressTextInput.getEditableText();
        if (editable != null) {
            emailAddress = editable.toString().trim();
        }
        if (emailAddress.isEmpty()) {
            binding.emailAddressLayout.setError(getString(R.string.email_address_is_required));
            binding.emailAddressTextInput.requestFocus();
            return;
        }

        final int passwordMinLength = getResources().getInteger(R.integer.password_min_length);
        String password = "";
        editable = binding.passwordTextInput.getEditableText();
        if (editable != null) {
            password = editable.toString();
        }
        if (password.isEmpty()) {
            binding.passwordLayout.setError(getString(R.string.password_is_required));
            binding.passwordTextInput.requestFocus();
            return;
        } else if (password.length() < passwordMinLength) {
            binding.passwordLayout.setError(getString(R.string.password_too_short));
            binding.passwordTextInput.requestFocus();
            return;
        }

        String verify = "";
        editable = binding.verifyPasswordTextInput.getEditableText();
        if (editable != null) {
            verify = editable.toString();
        }
        if (!verify.equals(password)) {
            binding.verifyPasswordLayout.setError(getString(R.string.passwords_must_match));
            binding.verifyPasswordTextInput.requestFocus();
            return;
        }

        updateUI(true);
        viewModel.register(emailAddress, password, this::registrationHandler);
    }

    private void registrationHandler(String error) {
        updateUI(false);
        if (error == null) {
            error = "";
        }
        error = error.trim();
        if (error.isEmpty()) {
            open(this, VerifyEmailActivity.class, true);
        } else {
            showModalAlert(this, getString(R.string.oops), error);
        }
    }

    private void updateUI(boolean locked) {
        binding.progressHorizontal.setVisibility(locked ? View.VISIBLE : View.INVISIBLE);
        binding.emailAddressLayout.setEnabled(!locked);
        binding.passwordLayout.setEnabled(!locked);
        binding.verifyPasswordLayout.setEnabled(!locked);
        binding.registerButton.setEnabled(!locked);
    }
}

File: app/src/main/java/de/eloc/eloc_control_panel/ng3/activities/RegisterActivity.kt
package de.eloc.eloc_control_panel.ng3.activities

import android.os.Bundle
import android.view.View
import androidx.lifecycle.ViewModelProvider
import de.eloc.eloc_control_panel.R
import de.eloc.eloc_control_panel.databinding.ActivityRegisterBinding
import de.eloc.eloc_control_panel.ng2.activities.TextInputWatcher
import de.eloc.eloc_control_panel.ng3.data.UserAccountViewModel

class RegisterActivity : ThemableActivity() {
    private lateinit var binding: ActivityRegisterBinding
    private lateinit var viewModel: UserAccountViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityRegisterBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModel = ViewModelProvider(this)[UserAccountViewModel::class.java]

        setListeners()
        setWatchers()
        updateUI(false)
    }

    private fun setWatchers() {
        binding.emailAddressTextInput.addTextChangedListener(TextInputWatcher(binding.emailAddressLayout))
        binding.passwordTextInput.addTextChangedListener(TextInputWatcher(binding.passwordLayout))
        binding.verifyPasswordTextInput.addTextChangedListener(TextInputWatcher(binding.verifyPasswordLayout))
    }

    private fun setListeners() {
        binding.backButton.setOnClickListener { openLogin() }
        binding.loginButton.setOnClickListener { openLogin() }
        binding.registerButton.setOnClickListener { register() }
        binding.root.setOnClickListener { hideKeyboard() }
        overrideGoBack { openLogin() }
    }

    private fun openLogin() = open(LoginActivity::class.java, true)

    private fun register() {
        hideKeyboard()
        binding.emailAddressLayout.error = null
        binding.passwordLayout.error = null
        binding.verifyPasswordLayout.error = null
        val emailAddress = binding.emailAddressTextInput.editableText?.toString()?.trim() ?: ""
        if (emailAddress.isEmpty()) {
            binding.emailAddressLayout.error = getString(R.string.email_address_is_required)
            binding.emailAddressTextInput.requestFocus()
            return
        }

        val passwordMinLength = resources.getInteger(R.integer.password_min_length)
        val password = binding.passwordTextInput.editableText?.toString() ?: ""
        if (password.isEmpty()) {
            binding.passwordLayout.error = getString(R.string.password_is_required)
            binding.passwordTextInput.requestFocus()
            return
        } else if (password.length < passwordMinLength) {
            binding.passwordLayout.error = getString(R.string.password_too_short)
            binding.passwordTextInput.requestFocus()
            return
        }

        val verify = binding.verifyPasswordTextInput.editableText?.toString() ?: ""
        if (verify != password) {
            binding.verifyPasswordLayout.error = getString(R.string.passwords_must_match)
            binding.verifyPasswordTextInput.requestFocus()
            return
        }

        updateUI(true)
        viewModel.register(emailAddress, password, this::registrationHandler)
    }

    private fun registrationHandler(error: String) {
        updateUI(false)
        val err = error.trim()
        if (err.isEmpty()) {
            open(VerifyEmailActivity::class.java, true)
        } else {
            showModalAlert(getString(R.string.oops), err)
        }
    }

    private fun updateUI(locked: Boolean) {
        binding.progressHorizontal.visibility = if (locked) View.VISIBLE else View.INVISIBLE
        binding.emailAddressLayout.isEnabled = !locked
        binding.passwordLayout.isEnabled = !locked
        binding.verifyPasswordLayout.isEnabled = !locked
        binding.registerButton.isEnabled = !locked
    }
}